module/zfs/dmu.c

```diff
@@ -1433,6 +1477,114 @@ dmu_write_uio(objset_t *os, uint64_t object, zfs_uio_t *uio, uint64_t size,
 }
 #endif /* _KERNEL */
 
+static void
+dbuf_cached_bps(spa_t *spa, blkptr_t *bps, uint_t nbps,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	int cached_flags;
+
+	if (bps == NULL)
+		return;
+
+	for (size_t blk_off = 0; blk_off < nbps; blk_off++) {
+		blkptr_t *bp = &bps[blk_off];
+		uint64_t *u64p;
+
+		if (BP_IS_HOLE(bp))
+			continue;
+
+		cached_flags = arc_cached(spa, bp);
+		if (cached_flags == 0)
+			continue;
+
+		u64p = (cached_flags & ARC_CACHED_IN_L2) ? l2sz : l1sz;
+		*u64p += BP_GET_LSIZE(bp);
+	}
+}
+
+/*
+ * Estimate DMU object cached size.
+ */
+int
+dmu_object_cached_size(objset_t *os, uint64_t object,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	dnode_t *dn;
+	dmu_object_info_t doi;
+	int level = 1;
+	int err = 0;
+
+	*l1sz = *l2sz = 0;
+
+	if (dnode_hold(os, object, FTAG, &dn) != 0)
+		return (0);
+
+	dmu_object_info_from_dnode(dn, &doi);
+
+	for (uint64_t off = 0; off < doi.doi_max_offset;
+	    off += dmu_prefetch_max) {
+		/* dbuf_read doesn't prefetch L1 blocks. */
+		dmu_prefetch(os, object, level, off,
+		    dmu_prefetch_max, ZIO_PRIORITY_NOW);
```

Comment:
Since #15076 dmu_prefetch_max limits not the amount of L0 data, but the amount of data/indirects to prefetch on the level requested, so since we use level = 1, this loop could go in bigger steps.  Though not a huge deal.

Also in #15951 I am adding dmu_prefetch_by_dnode().  Could be a small optimization to use it.

Reply:
Switched to use dmu_prefetch_by_dnode()